File: dnnmobile-app-android/app/src/main/java/de/hfabi/dnnmobile/benchmark/ModelConfig.kt
package de.hfabi.dnnmobile.benchmark

/**
 * Model Config
 *
 * Description of data sets and corresponding models.
 */
object ModelConfig {

    private const val dataDir = "datasets"
    private const val tfModelDir = "tensorflow"
    private const val ptModelDir = "pytorch"

    open class Dataset(
        var name: String,
        val trainInput: String,
        val trainTarget: String,
        val testInput: String,
        val testTarget: String,
        val featuresInput: Int,
        val featuresTarget: Int,
        val trainSampleCount: Int,
        val testSampleCount: Int
    )

    object FashionMnistSmall : Dataset(
        name = "Fashion Mnist Small",
        trainInput = "$dataDir/fashion_mnist_600_train_input.csv",
        trainTarget = "$dataDir/fashion_mnist_600_train_target_categorical.csv",
        testInput = "$dataDir/fashion_mnist_100_test_input.csv",
        testTarget = "$dataDir/fashion_mnist_100_test_target_categorical.csv",
        featuresInput = 784,
        featuresTarget = 10,
        trainSampleCount = 600,
        testSampleCount = 100
    )

    object BreastCancerWisconsin : Dataset(
        name = "Breast Cancer Wisconsin",
        trainInput = "$dataDir/breastcancer_train_input.csv",
        trainTarget = "$dataDir/breastcancer_train_target.csv",
        testInput = "$dataDir/breastcancer_test_input.csv",
        testTarget = "$dataDir/breastcancer_test_target.csv",
        featuresInput = 30,
        featuresTarget = 1,
        trainSampleCount = 398,
        testSampleCount = 171
    )

    object Mnist : Dataset(
        name = "Mnist",
        trainInput = "$dataDir/mnist_train_input.csv",
        trainTarget = "$dataDir/mnist_train_target_categorical.csv",
        testInput = "$dataDir/mnist_test_input.csv",
        testTarget = "$dataDir/mnist_test_target_categorical.csv",
        featuresInput = 784,
        featuresTarget = 10,
        trainSampleCount = 60000,
        testSampleCount = 10000
    )

    object MnistSmall : Dataset(
        name = "Mnist",
        trainInput = "$dataDir/mnist_600_train_input.csv",
        trainTarget = "$dataDir/mnist_600_train_target_categorical.csv",
        testInput = "$dataDir/mnist_100_test_input.csv",
        testTarget = "$dataDir/mnist_100_test_target_categorical.csv",
        featuresInput = Mnist.featuresInput,
        featuresTarget = Mnist.featuresTarget,
        trainSampleCount = 600,
        testSampleCount = 100
    )

    object Model {
        object TensorFlow {
            const val BreastCancerWisconsin = "$tfModelDir/breast_cancer_wisconsin.tflite"
            const val FashionMnist = "$tfModelDir/fashion_mnist.tflite"
            const val Mnist = "$tfModelDir/mnist.tflite"
        }

        object Pytorch {
            const val BreastCancerWisconsin = "$ptModelDir/breast_cancer_wisconsin.ptl"
            const val FashionMnist = "$ptModelDir/fashion_mnist.ptl"
            const val Mnist = "$ptModelDir/mnist.ptl"
        }
    }
}

File: dnnmobile-app-ios/dnnmobile/benchmark/ModelConfig.swift
//
//  ModelConfig.swift
//  dnnmobile
//
//  Created by Fabian on 10.03.23.
//

import Foundation

protocol DnnDataset {
    var name: String { get }
    var trainInput: String { get }
    var trainTarget: String { get }
    var testInput: String { get }
    var testTarget: String { get }
    var featuresInput: Int { get }
    var featuresTarget: Int { get }
    var trainSampleCount: Int { get }
    var testSampleCount: Int { get }
}

protocol DnnModel {
    var bundleName: String { get }
    var fileName: String { get }
}

protocol CoreMlDnnModel: DnnModel {
    var featureLabelInput: String { get }
    var featureLabelOutput: String { get }
    var featureLabelOutputClass: String { get }
    var allowedEpochs: Array<Int> { get }
}

extension CoreMlDnnModel {
    var fileName: String {
        get {
            return "\(bundleName).mlmodelc"
        }
    }
}

struct ModelConfig {
    
    private static let dataDir = "datasets"
    
    struct FashionMnistSmall : DnnDataset {
        let name = "Fashion Mnist Small"
        let trainInput = "\(dataDir)/fashion_mnist_600_train_input"
        let trainTarget = "\(dataDir)/fashion_mnist_600_train_target_categorical"
        let testInput = "\(dataDir)/fashion_mnist_100_test_input"
        let testTarget = "\(dataDir)/fashion_mnist_100_test_target_categorical"
        let featuresInput = 784
        let featuresTarget = 10
        let trainSampleCount = 600
        let testSampleCount = 100
    }
    
    struct BreastCancerWisconsin : DnnDataset {
        let name = "Breast Cancer Wisconsin"
        let trainInput = "\(dataDir)/fashion_mnist_600_train_input"
        let trainTarget = "\(dataDir)/fashion_mnist_600_train_target_categorical"
        let testInput = "\(dataDir)/fashion_mnist_100_test_input"
        let testTarget = "\(dataDir)/fashion_mnist_100_test_target_categorical"
        let featuresInput = 30
        let featuresTarget = 1
        let trainSampleCount = 398
        let testSampleCount = 171
    }
    
    struct Mnist : DnnDataset {
        let name = "Mnist"
        let trainInput = "\(dataDir)/fashion_mnist_600_train_input"
        let trainTarget = "\(dataDir)/fashion_mnist_600_train_target_categorical"
        let testInput = "\(dataDir)/fashion_mnist_100_test_input"
        let testTarget = "\(dataDir)/fashion_mnist_100_test_target_categorical"
        let featuresInput = 784
        let featuresTarget = 10
        let trainSampleCount = 60000
        let testSampleCount = 10000
    }
    
    struct MnistSmall : DnnDataset {
        let name = "Mnist"
        let trainInput = "\(dataDir)/fashion_mnist_600_train_input"
        let trainTarget = "\(dataDir)/fashion_mnist_600_train_target_categorical"
        let testInput = "\(dataDir)/fashion_mnist_100_test_input"
        let testTarget = "\(dataDir)/fashion_mnist_100_test_target_categorical"
        let featuresInput = 784
        let featuresTarget = 10
        let trainSampleCount = 600
        let testSampleCount = 100
    }
    
    struct CoreMlMnist: CoreMlDnnModel {
        var featureLabelInput = "ip_1_input"
        var featureLabelOutput = "ip_6_output"
        var featureLabelOutputClass = "ip_6_output_true"
        var allowedEpochs = [50]
        let bundleName = "MNIST100-500Updatable"
    }
}
